pitop/pma/ultrasonic_sensor.py

```diff
@@ -1,61 +1,47 @@
-from gpiozero import SmoothedInputDevice
-from threading import Event, Lock
-
-from pitopcommon.logger import PTLogger
-
 from pitop.core.mixins import (
     Stateful,
     Recreatable,
 )
-from pitop.pma.common import get_pin_for_port
-
-
-# Modified version of gpiozero's DistanceSensor class that only uses 1 pin
-#
-# Note: all private member variables are semi-private to follow upstream gpiozero convention
-# and to override inherited functions
-class UltrasonicSensor(Stateful, Recreatable, SmoothedInputDevice):
-    ECHO_LOCK = Lock()
-
-    def __init__(
-        self,
-        port_name,
-        queue_len=9,
-        max_distance=3,
-        threshold_distance=0.3,
-        partial=False,
-        name="ultrasonic"
-    ):
+from pitop.pma.common.utils import Port
+from .ultrasonic_sensor_base import UltrasonicSensorMCU, UltrasonicSensorRPI
+
+
+valid_analog_ports = ["A1", "A3"]
+
+
+class UltrasonicSensor(Stateful, Recreatable):
+    def __init__(self,
+                 port_name,
+                 queue_len=5,
+                 max_distance=3,
+                 threshold_distance=0.3,
+                 partial=False,
+                 name="ultrasonic"
+                 ):
 
+        assert (port_name in Port)
         self._pma_port = port_name
         self.name = name
 
-        SmoothedInputDevice.__init__(self,
-                                     get_pin_for_port(self._pma_port),
-                                     pull_up=False,
-                                     queue_len=queue_len,
-                                     sample_wait=0.06,
-                                     partial=partial,
-                                     ignore=frozenset({None}),
-                                     )
-
-        try:
-            if max_distance <= 0:
-                raise ValueError('invalid maximum distance (must be positive)')
-            self._max_distance = max_distance
-            self.threshold = threshold_distance / max_distance
-            self.speed_of_sound = 343.26  # m/s
-
-            self._echo = Event()
-            self._echo_rise = None
-            self._echo_fall = None
-            self.pin.edges = 'both'
-            self.pin.bounce = None
-            self.pin.when_changed = self._echo_changed
-            self._queue.start()
-        except Exception:
-            self.close()
-            raise
+        if port_name in valid_analog_ports:
```

Comment:
I think there should be a more explicit api for selecting these implementations than passing a particular port number, along the lines of CameraType. Also this blocks using ports A1 or A3 for UltrasonicSensorRPI which I don't think is necessary?

Reply:
UltrasonicSensorRPi = controlled via RPi digital pins
UltrasonicSensorMCU = controlled via expansion plate MCU pins and accessed over I2C by Raspberry Pi

UltrasonicSensorRPi can't use A1 and A3 so I don't think there is a need for a user to choose an implementation through an API as it depends solely on which port they have it connected to. It would be nice if the pros and cons of each were described, but in general, I think for all robotics Further content we should be using UltrasonicSensorMCU since it's more robust (which means telling user to plug into A1 or A3)

Reply:
Right yeah, confused myself. In that case this does seem like a good way of selecting the implementation, and we'll have to explain the differences and constraints in the docs or elsewhere.